### ISSUE TITLE
Lifecycle Practices

### ISSUE THREAD

**Issue by bonbonio** [zeit/next.js#9175]
Hello,

Forgive me, but I think the question might be a little wide-reaching for StackOverflow or perhaps Quora and I did try asking on Spectrum chat.

I cannot quite seem a use case with best practice advise or code examples for a web application like so; imagine I am building IMDB

I have an API which has all the movie data, this is in Hapi and is completed. I was using react-boilerplate to query the API and render it but for the supposed SEO benefits I want to use NextJS.

I have all the routes sorted in NextJS, and the site generally works; but I was consuming the API in pages and passing it to components like so;

```
MoviePage.getInitialProps = async function({ res, query }) {

  const req = await axios.get(`/city/${query.cityId}`);

  const { data } = req;

  return {
    data
  };
};
```

But when I refreshed the page (whilst running `npm run start`) the page would break; so I read that getInitialProps is only called when its running server-side.

So now I have this; a very standard react class with componentDidMount, state and render.

```
class MoviePage extends React.Component {
  state = {};

  loadByMovieSlug = () => {
  }
  
  componentDidMount() {
    this.loadByMovieSlug();
  }

  render() { etc}
}
```

But for my clarification; I have the following questions;

- Would using NextJS to fetch API data in componentDidMount still provide an SEO benefit; I believe these pages would now no longer be SSR - a key benefit of this framework. Is this correct?
- How do I access route params inside these components (or child components if I don't want to pass them down as props) if I must have my pages like the componentDidMount example?
- Is there meant to be a way to combine getInitialProps and componentDidMount into one page class; or is there only one I should be using.
- Is there an example anywhere of a page loading data from an API and passing it down to components that is best practice in terms of code and also SEO? I cannot find one on GitHub or Google... 

I would happily write extra documentation on these topics should it be of value once its clarified to me - or am I the only one?

**Comment by rafaelalmeidatk** [comment]
> so I read that getInitialProps is only called when its running server-side.

This is not true, the method is called both on the server and on client-side route changes. Your page is breaking for another reason that is hard to tell because you didn't provide any error.

I would not do what you did in the last snippet, there is no reason to 1) use a class 2) use derived state from props (it is literally the same object duplicated in two places).

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
